framework/src/meshgenerators/MeshGenerator.C

```diff
@@ -20,15 +20,39 @@ MeshGenerator::validParams()
                         false,
                         "Whether or not to show mesh info after generating the mesh "
                         "(bounding box, element types, sidesets, nodesets, subdomains, etc)");
+  params.addParam<std::vector<std::string>>(
+      "selected_mesh_metadata_to_retain",
+      std::vector<std::string>(),
+      "Names of input mesh's metadata that are selected to retain.");
+  params.addParam<bool>("retain_all_input_mesh_metadata",
+                        false,
+                        "Whether to retain all the mesh metadata of the input mesh.");
+
+  params.addParamNamesToGroup("selected_mesh_metadata_to_retain retain_all_input_mesh_metadata",
+                              "Mesh Metadata Retainment");
 
   params.registerBase("MeshGenerator");
 
   return params;
 }
 
 MeshGenerator::MeshGenerator(const InputParameters & parameters)
-  : MooseObject(parameters), MeshMetaDataInterface(this), _mesh(_app.actionWarehouse().mesh())
+  : MooseObject(parameters),
+    MeshMetaDataInterface(this),
+    _mesh(_app.actionWarehouse().mesh()),
+    _selected_mesh_metadata_to_retain(
+        getParam<std::vector<std::string>>("selected_mesh_metadata_to_retain")),
+    _retain_all_input_mesh_metadata(getParam<bool>("retain_all_input_mesh_metadata"))
 {
+  if (!isParamValid("input"))
+  {
+    if (_retain_all_input_mesh_metadata)
+      paramError("retain_all_input_mesh_metadata",
+                 "In the absence of an input mesh, this parameter must not be true.");
+    if (!_selected_mesh_metadata_to_retain.empty())
+      paramError("selected_mesh_metadata_to_retain",
+                 "In the absence of an input mesh, this parameter must be empty.");
+  }
```

Comment:
I think this can be done from the inerface constructor, if you pass the parameters